test/Main.hs

```diff
@@ -20,12 +25,29 @@ import           Test.Tasty.QuickCheck
 
 import           Crypto.Data.Auth.Tree (Tree)
 import qualified Crypto.Data.Auth.Tree as Tree
+import qualified Crypto.Data.Auth.Tree.Cryptonite as Cryptonite
+import           Crypto.Data.Auth.Tree.Internal
 import qualified Crypto.Data.Auth.Tree.Proof as Tree
-import           Crypto.Hash (SHA256)
+import qualified Crypto.Hash as Cryptonite
+
+import qualified Data.Digest.XXHash as Mock
 
 type Key = Word8
 type Val = Word8
 
+type CryptoniteDigest = Cryptonite.Digest Cryptonite.SHA256
+type MockDigest      = Mock.XXHash
+
+instance MerkleHash (Cryptonite.Digest Cryptonite.SHA256) where
+    emptyHash = Cryptonite.emptyHash
+    hashLeaf  = Cryptonite.hashLeaf
+    hashNode  = Cryptonite.hashNode
+
+instance MerkleHash MockDigest where
+    emptyHash      = minBound
+    hashLeaf k v   = Mock.xxHash' (convert k <> convert v)
+    hashNode d1 d2 = d1 + d2
```

Comment:
What does `+` do here?

Reply:
It's the _tentative_ translation of the cryptonite implementation:

```
hashNode
    :: HashAlgorithm a
    => Digest a -> Digest a -> Digest a
hashNode l r =
    hashFinalize $ flip hashUpdates [l, r]
                 $ flip hashUpdate (BS.singleton 1)
                 $ hashInit
```

In the cryptonite world we had all that machinery to call `hashUpdate` which takes a list of bytearrays and hashes them together. Here, however, `d1` and `d2` are of type `XXHash`, which is just a wrapper around `Word32`. Therefore, the only meaningful thing I could do was to sum them together. I suspect it shouldn't matter _too much_ for our mock implementation, but I am open to different suggestions. Xor them maybe? Bit twiddling?

Reply:
Ah I see, it's actually summing - I think since Word32 just wraps on overflow it's fine.